File: ex4/ex4.c
#include <stdio.h>
#include <stdlib.h>

void swap(int *arr, int a, int b) {
    int temp = arr[a];
    arr[a] = arr[b];
    arr[b] = temp;
}

int countRounds(int *arr, int n) {
    int rounds = 1;
    for (int i = 1; i < n; i++) {
        if (arr[i] < arr[i - 1]) {
            rounds++;
        }
    }
    return rounds;
}

int main() {
    int n, m;
    scanf("%d %d", &n, &m);

    int *arr = (int *)malloc(n * sizeof(int));
    for (int i = 0; i < n; i++) {
        scanf("%d", &arr[i]);
    }

    int *positions = (int *)malloc((n + 1) * sizeof(int));
    for (int i = 0; i < n; i++) {
        positions[arr[i]] = i;
    }

    int rounds = countRounds(arr, n);
    printf("%d\n", rounds);

    for (int i = 0; i < m; i++) {
        int a, b;
        scanf("%d %d", &a, &b);
        a--; b--;

        swap(arr, a, b);
        positions[arr[a]] = a;
        positions[arr[b]] = b;

        rounds = countRounds(arr, n);
        printf("%d\n", rounds);
    }

    free(arr);
    free(positions);

    return 0;
}

File: ex4/ex4.py
def swap(arr, a, b):
    arr[a], arr[b] = arr[b], arr[a]

def count_rounds(arr):
    rounds = 1
    for i in range(1, len(arr)):
        if arr[i] < arr[i - 1]:
            rounds += 1
    return rounds

def main():
    n, m = map(int, input().split())
    arr = list(map(int, input().split()))

    positions = [0] * (n + 1)
    for i in range(n):
        positions[arr[i]] = i

    rounds = count_rounds(arr)
    print(rounds)

    for _ in range(m):
        a, b = map(int, input().split())
        a -= 1
        b -= 1

        swap(arr, a, b)
        positions[arr[a]] = a
        positions[arr[b]] = b

        rounds = count_rounds(arr)
        print(rounds)

if __name__ == "__main__":
    main()